# install.packages('LDlinkR') # https://ldlink.nci.nih.gov/?tab=snpclip
#install.packages('sjmisc')
library(LDlinkR)
setwd("/Users/mtomason/Documents/projects/retina/data_UKBiobank/03_gwas/hits_extract")

# INSTRUCTIONS 
# - take [pheno]__topHits file generated by hit_to_csv.R
# OUTPUT
# - an LD thinned subset of [pheno]_topHits based on R2+distance

# INITIALIZE APPROPRIATELY ####################################################

# OLD AV CLASSIFICATION GWAS (AVUncertain)
###hits_file_all <- "2020_08_12__62751_resid_qqnorm_D9/GWAS_results_pruned_R2_0.01_GBR.csv"
###hits_file_artery <- "2020_08_16__115820_rsid_qqnorm_D9_artery/GWAS_results_pruned_R2_0.01_GBR_artery.csv"
###hits_file_vein <- "2020_08_16__115931_rsid_qqnorm_D9_vein/GWAS_results_pruned_R2_0.01_GBR_vein.csv"
# fill in GWAS-specific p-values from raw BGENIE output
###artery_files_path <- "/Users/mtomason/Documents/projects/retina/data_UKBiobank/03_gwas/qqplot/2020_08_16__115820_rsid_qqnorm_D9_artery/_input/"
###vein_files_path <- "/Users/mtomason/Documents/projects/retina/data_UKBiobank/03_gwas/qqplot/2020_08_16__115931_rsid_qqnorm_D9_vein/_input/"
###all_files_path <- "/Users/mtomason/Documents/projects/retina/data_UKBiobank/03_gwas/qqplot/2020_08_12__62751_resid_qqnorm_D9/_input/"

# LWNet threshold 0.00
hits_file_all <- "2020_10_03__62751_trait_qqnorm/GWAS_results_pruned_R2_0.1_GBR_all.csv"
hits_file_artery <- "2020_10_03__62751_trait_qqnorm_artery00/GWAS_results_pruned_R2_0.1_GBR_artery.csv"
hits_file_vein <- "2020_10_03__62751_trait_qqnorm_vein00/GWAS_results_pruned_R2_0.1_GBR_vein.csv"
all_files_path <- "/Users/mtomason/Documents/projects/retina/data_UKBiobank/03_gwas/qqplot/2020_10_03__62751_trait_qqnorm/_input/"
artery_files_path <- "/Users/mtomason/Documents/projects/retina/data_UKBiobank/03_gwas/qqplot/2020_10_03__62751_trait_qqnorm_artery00/_input/"
vein_files_path <- "/Users/mtomason/Documents/projects/retina/data_UKBiobank/03_gwas/qqplot/2020_10_03__62751_trait_qqnorm_vein00/_input/"

pop <- "GBR" #"GBR" "CEU"
r2_thresh <- "0.1"
distance_thresh <- 500000
# INITIALIZE APPROPRIATELY ####################################################

# read hits
start.time <- Sys.time()
hits_all <- read.csv(hits_file_all, header=TRUE,check.names=FALSE)
hits_artery <- read.csv(hits_file_artery, header=TRUE,check.names=FALSE)
hits_vein <- read.csv(hits_file_vein, header=TRUE,check.names=FALSE)
# add column to identify vessel type
hits_all$GWAS_type <- rep(c("combined"), each=nrow(hits_all))
hits_all <- hits_all[order(hits_all$median_tortuosity.log10p, decreasing=TRUE), ] # order by pvalue
hits_artery$GWAS_type <- rep(c("artery"), each=nrow(hits_artery)) 
hits_artery <- hits_artery[order(hits_artery$median_tortuosity.log10p, decreasing=TRUE), ] # order by pvalue
hits_vein$GWAS_type <- rep(c("vein"), each=nrow(hits_vein)) 
hits_vein <- hits_vein[order(hits_vein$median_tortuosity.log10p, decreasing=TRUE), ] # order by pvalue

# OPTION 3 # COMBINED HITS (manually replace "all")
hits_AV <- rbind(hits_vein,hits_artery)
hits_AV <- hits_AV[order(hits_AV$median_tortuosity.log10p, decreasing=TRUE), ] # order by pvalue
hits <- rbind(hits_AV,hits_all) 

hits["pruned"] = FALSE # add column to keep track of pruned SNPs

#########################i <- 1; while (i <= nrow(hits)-1) { # check each SNP ################## WHY -1?????????
i <- 1; while (i <= nrow(hits)) { # check each SNP
  snp_i <- hits[i,] 
  if(snp_i$pruned == FALSE){ # if it has not been pruned...
    j <- i+1; while (j <= nrow(hits)) { # ...then look for SNPs to prune among all other SNPs...
      snp_j <- hits[j,]
      if(snp_i$chr == snp_j$chr){ # ...which are in the same chr...
        if(abs(snp_i$pos - snp_j$pos) < distance_thresh){ # ...and which are less than 500Kb away...
          if(snp_j$pruned == FALSE){ # ...and which have not been pruned...  
            write(paste(snp_i$rsid, snp_j$rsid), stdout())
            error = tryCatch({
              LDLinkOutput <- LDpair(snp_i$rsid, snp_j$rsid, pop, token = "f0c92fa3ea9b", file = FALSE)
              if(LDLinkOutput$r2 > r2_thresh){ # ...if not in LD...
                hits[j,"pruned"] <- TRUE # ...then prune it away
              }
            }, error = function(error_condition) {error = TRUE}) # prune also in case of error
            if(!is.null(error) && error == TRUE){ 
              hits[j,"pruned"] <- TRUE 
            } 
          }
        }
      }
      j <- j+1  
    }
  }
  i <- i+1
}
result <- hits[hits["pruned"]==FALSE,] # only variants that were not pruned

################################################################################################################
# add columns to hold the corresponding GWAS-specific pvalues
result["signal_type"] = "undefined"
result["log10p.vein-GWAS"] = -1
result["log10p.artery-GWAS"] = -1
result["log10p.combined-GWAS"] = -1

# order by chromosome to avoid re-loading BGENIE output unnecessarily
result <- result[order(result$chr),]

chromo_in_last_iteration=-1
i <- 1; while (i <= nrow(result)) { # check each SNP
  print(paste("Adding GWAS-specific p-values: processing SNP",i,"of",nrow(result)))
  snp_i <- as.character(result[i,]$rsid)
  chr_i <- result[i,]$chr
  if (chromo_in_last_iteration != chr_i){
    # load BGENIE outputs
    artery_file <- paste0(artery_files_path, "output_ukb_imp_chr",chr_i,"_v3.txt")
    chromo_artery <- read.csv(artery_file, header=TRUE,check.names=FALSE, sep=" ")
    vein_file <- paste0(vein_files_path, "output_ukb_imp_chr",chr_i,"_v3.txt")
    chromo_vein <- read.csv(vein_file, header=TRUE,check.names=FALSE, sep=" ")
    all_file <- paste0(all_files_path, "output_ukb_imp_chr",chr_i,"_v3.txt")
    chromo_all <- read.csv(all_file, header=TRUE,check.names=FALSE, sep=" ")
  }
  # artery-GWAS
  result[i,]$`log10p.artery-GWAS` <- chromo_artery[chromo_artery$rsid==snp_i,][1,]$`median_tortuosity-log10p`
  # vein-GWAS 
  result[i,]$`log10p.vein-GWAS` <- chromo_vein[chromo_vein$rsid==snp_i,][1,]$`median_tortuosity-log10p`
  # combined-GWAS
  result[i,]$`log10p.combined-GWAS` <- chromo_all[chromo_all$rsid==snp_i,][1,]$`median_tortuosity-log10p`
  
  # increment loop vars
  i <- i+1
  chromo_in_last_iteration = chr_i
}

################################################################################################################
# update the signal type according to the GWAS-specific pvalues
###bonf_thr = -log10(0.05/nrow(result)) # correct for number of results
bonf_thr = 7.301 # Bonferroni = -log10(5E-8) = 7.301
i <- 1; while (i <= nrow(result)) { # check each SNP
  print(paste("Adding signal type: processing SNP",i,"of",nrow(result)))
  log10p_artery <- result[i,]$`log10p.artery-GWAS` 
  log10p_vein <- result[i,]$`log10p.vein-GWAS` 
  logg10p_combined <- result[i,]$`log10p.combined-GWAS`
  # artery-specific
  if (log10p_artery > bonf_thr && log10p_vein < bonf_thr) {
    result[i,]$signal_type <- "artery"
  }
  # vein-specific
  if (log10p_artery < bonf_thr && log10p_vein > bonf_thr) {
    result[i,]$signal_type <- "vein"
  }
  # both
  if (log10p_artery > bonf_thr && log10p_vein > bonf_thr) {
    result[i,]$signal_type <- "both"
  }
  # combined
  if (log10p_artery < bonf_thr && log10p_vein < bonf_thr) {
    result[i,]$signal_type <- "combined"
  }
  i <- i+1
}

################################################################################################################
# output thinned list of SNPs
result[,"pruned"] <- NULL # remove fields that are not needed in the output
result[,"GWAS_type"] <- NULL
result <- result[with(result, order(chr, pos, decreasing = FALSE)), ] # order by chrom and position
names(result)[names(result)=="median_tortuosity.log10p"] <- "log10p"
names(result)[names(result)=="median_tortuosity_beta"] <- "beta"
names(result)[names(result)=="median_tortuosity_se"] <- "se"
write.csv(result, paste0("GWAS_results_pruned_R2_", r2_thresh,"_",pop,"_merged_AV_resid_qqnorm.csv"), row.names = F)
end.time <- Sys.time()
time.taken <- end.time - start.time
write(paste("execution took", as.integer(time.taken),"seconds"), stdout())

